### ISSUE TITLE
Implement memoize and memoizeOnSuccess from Monix Task

### ISSUE THREAD

**Issue by Avasil** [monix/monix-bio#7]
General flow:
- Comment in this issue to let others know that someone is working on it
- Copy scaladoc and the implementation from [Monix Task](https://github.com/monix/monix) in similar place
- Modify signature to take advantage of typed errors if possible
- Implementation might require some changes, in most cases the compiler should guide you, also look for already ported operators for inspiration. The most common change is to use `BiCallback` instead of `Callback` to be able to signal errors on two different channels ("typed" channel and hidden channel for "fatal" errors).
- Copy the corresponding tests from Monix. Try to use `BIO` instead of `Task` with `E` different than `Throwable` when possible to uncover any possible bugs in the internals. 
- If you test for a failure, try to test for both "expected" failure (error from `E`, it will be `CancelableFuture(Left))` after `runToFuture`) and "fatal" failure (it will be a failed `CancelableFuture`).
- Take a look at the existing tests to see if there aren't any commented ones using newly implemented function and uncomment them.

Don't be shy to ping me if you need any advice or want to consult the solution. I really appreciate the involvement and I'm happy to help!

**Comment by SvenvDam** [comment]
I'll pick this one up!

**Comment by SvenvDam** [comment]
I feel like `memoizeOnSuccess` should cache both sucesses and expected errors, while `memoize` would also cache the unexpected/fatal errors.
In that case the name `memoizeOnSuccess` would be a bit misleading though I suppose, any suggestions for a better name?

**Comment by Avasil** [comment]
> I feel like memoizeOnSuccess should cache both sucesses and expected errors

Could you elaborate? I feel like we should always consider `E` as a failure

We could add something like `memoizeOnResult` but I'm not sure if it adds much value over standard `memoize`

**Comment by SvenvDam** [comment]
My question is the following:
In Monix Task, `memoizeOnSuccess` will only cache succesfull results and failures would be recalculated every time.

In our case, I'm not fully sure whether or not an expected failure should be cached. 
Given that this failure case is expected I guess it would be reasonable to assume the user will handle these failures gracefully and not caching it would result in possibly redundant computation.

**Comment by Avasil** [comment]
I would say it shouldn't be cached in `memoizeOnSuccess`. I'm not sure what was the original use case to implement it but I guess it was about a success of type `A`. 

We can still cache typed errors with `memoize` or with `memoizeOnSuccess.attempt`.
We could also use `memoizeOnSuccess` to make it memoize typed errors instead with `flip` (which is not yet implemented but we will have it).

If we consider efficiency then:
- if `memoizeOnSuccess` caches error then the user will still have to handle it each time it is called which is probably redundant in most cases.
- if it doesn't cache typed errors then we can do `task.onErrorHandleWith(handleError).memoizeOnSuccess` to cache it with already handled errors.

Basically I don't see a lot of value in such operator because I feel like it can be easily expressed with what we get with `memoize` + `memoizeOnSuccess` working on just `A`. Does it make sense or do you see some use cases that would it help with?

**Comment by SvenvDam** [comment]
Thanks for the explanation! I'll implement it accordingly. I'll probably need some time for this issue since there are quite a few internals that have to be ported over and the holidays coming up. But I'll continue to work on it! If I need any assistance I'll post questions here.
Thanks again!

**Comment by Avasil** [comment]
No worries, take your time. I think most of the code will stay the same but there might be some changes necessary to use `BiCallback` instead of `Callback`. There should be some examples in `monix-bio` for other operators. If you're stuck I can take a closer look what exactly needs to be done and explain in more detail :)
